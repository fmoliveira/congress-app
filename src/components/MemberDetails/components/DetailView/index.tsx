import React from "react"
import { Link } from "react-router-dom"
import styled from "styled-components"

import { RequestStatusType } from "../../../../reducers/requestStatus"

import { Map } from "../../../_common"
import { ICommittee, IMemberInfo } from "../../types"
import { Committee, News } from "../index"
import { ListHeader } from "./ListHeader"
import { MemberInfo } from "./MemberInfo"
import { SkeletonItem } from "./SkeletonItem"

import back from "./back.svg"

export const LinkWrapper = styled(Link)`
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  padding: 0 1em;
  text-decoration: none;
  color: #999999;

  &:hover {
    color: #555555;
  }
`

export const Icon = styled.img`
  width: auto;
  height: 16px;
`

export const LinkText = styled.span`
  margin: 0 1em;
`

export const Columns = styled.div`
  margin: 1em 0;
  display: flex;
  flex-direction: row;

  & > * {
    width: 50%:
    flex: 1;
    display: flex;
    flex-direction: column;
    height: 350px;
    overflow: hidden;
  }
`

interface IProps {
  committees: ICommittee[]
  info: IMemberInfo
  memberId: string
  office: string
  rssUrl: string
  status: RequestStatusType
  subcommittees: ICommittee[]
}

const DetailView = ({
  committees,
  info,
  memberId,
  office,
  rssUrl,
  status,
  subcommittees
}: IProps) => (
  <div>
    <header>
      <LinkWrapper to="/">
        <Icon src={back} alt="Go back to members list" />
        <LinkText>Back to members list</LinkText>
      </LinkWrapper>
    </header>

    <ListHeader />
    {status !== RequestStatusType.Success && <SkeletonItem />}
    {status === RequestStatusType.Success && <MemberInfo {...info} />}

    <Columns>
      <News memberId={memberId} feedUrl={rssUrl} />
      <Map address={office} />
    </Columns>

    <Committee title="Committees" items={committees} />
    <Committee title="Sub-committees" items={subcommittees} />
  </div>
)

export default DetailView
