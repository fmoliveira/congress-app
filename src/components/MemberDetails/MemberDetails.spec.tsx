import React from "react"
import { MemoryRouter } from "react-router"
import { cleanup, render } from "react-testing-library"

import DetailView from "./components/DetailView"
import { UnconnectedMemberDetails as MemberDetails } from "./index"

jest.mock("./components/DetailView", () => jest.fn(() => <div>DetailView</div>))

const setup = (props?: any) =>
  render(
    <MemoryRouter>
      <MemberDetails {...props} />
    </MemoryRouter>
  )

describe("MemberDetails", () => {
  afterEach(() => {
    cleanup()
    jest.clearAllMocks()
  })

  it("Should render without crashing", () => {
    const { container } = setup()
    expect(container.firstChild).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it("Should call action to get member details", () => {
    const getMemberDetails = jest.fn()
    const match = { params: { memberId: "M123" } }
    const props = { getMemberDetails, match }
    setup(props)

    expect(getMemberDetails).toHaveBeenCalledTimes(1)
    expect(getMemberDetails).toHaveBeenCalledWith("M123")
  })

  it("Should render detail view", () => {
    const props = {
      details: {
        currentParty: "I",
        facebookAccount: "",
        firstName: "Jack",
        gender: "male",
        lastName: "Sparrow",
        memberId: "M125", // we want the id to be taken from here, not from match param below
        roles: [
          {},
          {
            committees: [{ code: "COM1", name: "Committee One" }],
            office: "116 Fancy Skyscrapper",
            state: "XYZ",
            subcommittees: [{ code: "SCM2", name: "Sub-committee Two" }]
          }
        ],
        rssUrl: "https://loremipsum.com/news.rss",
        twitterAccount: "DangerousPirate",
        youtubeAccount: "MaroonedInNassau"
      },
      info: {},
      match: { params: { memberId: "M123" } }, // must differ from memberId above
      status: "SUCCESS"
    }
    setup(props)

    expect(DetailView).toHaveBeenCalledTimes(1)
    expect(DetailView).toBeCalledWith(
      {
        committees: [{ code: "COM1", name: "Committee One" }],
        info: {
          facebookAccount: "",
          firstName: "Jack",
          gender: "male",
          id: "M125",
          lastName: "Sparrow",
          nextElection: "",
          office: "116 Fancy Skyscrapper",
          party: "I",
          rssUrl: "https://loremipsum.com/news.rss",
          state: "XYZ",
          twitterAccount: "DangerousPirate",
          youtubeAccount: "MaroonedInNassau"
        },
        memberId: "M125",
        office: "116 Fancy Skyscrapper",
        rssUrl: "https://loremipsum.com/news.rss",
        status: "SUCCESS",
        subcommittees: [{ code: "SCM2", name: "Sub-committee Two" }]
      },
      {}
    )
  })

  it("Should supply current data from list view while loading", () => {
    const props = {
      info: {
        facebookAccount: "",
        firstName: "Jack",
        gender: "male",
        id: "M125", // we want the id to be taken from here, not from match param below
        lastName: "Sparrow",
        nextElection: "2045",
        office: "116 Fancy Skyscrapper",
        party: "I",
        rssUrl: "https://loremipsum.com/news.rss",
        state: "XYZ",
        twitterAccount: "DangerousPirate",
        youtubeAccount: "MaroonedInNassau"
      },
      match: { params: { memberId: "M123" } }, // must differ from id above
      status: "LOADING"
    }
    setup(props)

    expect(DetailView).toHaveBeenCalledTimes(1)
    expect(DetailView).toBeCalledWith(
      {
        committees: [],
        info: {
          facebookAccount: "",
          firstName: "Jack",
          gender: "male",
          id: "M125",
          lastName: "Sparrow",
          nextElection: "2045",
          office: "116 Fancy Skyscrapper",
          party: "I",
          rssUrl: "https://loremipsum.com/news.rss",
          state: "XYZ",
          twitterAccount: "DangerousPirate",
          youtubeAccount: "MaroonedInNassau"
        },
        memberId: "M125",
        office: "116 Fancy Skyscrapper",
        rssUrl: "https://loremipsum.com/news.rss",
        status: "LOADING",
        subcommittees: []
      },
      {}
    )
  })
})
