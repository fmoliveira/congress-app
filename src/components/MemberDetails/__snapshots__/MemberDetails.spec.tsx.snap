// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberDetails Should render without crashing 1`] = `
<div>
  <div>
    DetailView
  </div>
</div>
`;
